server/datastore/mysql/software.go

```diff
@@ -2251,459 +2252,689 @@ func (ds *Datastore) ListCVEs(ctx context.Context, maxAge time.Duration) ([]flee
+	VPPAppAdamIDList      *string `db:"vpp_app_adam_id_list"`
+	VPPAppVersionList     *string `db:"vpp_app_version_list"`
+	VPPAppPlatformList    *string `db:"vpp_app_platform_list"`
+	VPPAppIconUrlList     *string `db:"vpp_app_icon_url_list"`
+}
+
+func hostInstalledSoftware(ds *Datastore, ctx context.Context, hostID uint) ([]*hostSoftware, error) {
+	installedSoftwareStmt := `
+		SELECT
+			software_titles.id AS id,
+			host_software.software_id AS software_id,
+			host_software.last_opened_at,
+			NULL AS status
+		FROM 
+			host_software
+		INNER JOIN
+			software ON host_software.software_id = software.id
+		INNER JOIN
+			software_titles ON software.title_id = software_titles.id
+		WHERE
+			host_software.host_id = :host_id
+	`
+	installedSoftwareStmt, args, err := sqlx.Named(installedSoftwareStmt, map[string]any{
+		"host_id": hostID,
+	})
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "build named query for host installed software")
 	}
 
-	var onlySelfServiceClause string
-	if opts.SelfServiceOnly {
-		onlySelfServiceClause = ` AND ( si.self_service = 1 OR ( vat.self_service = 1 AND :is_mdm_enrolled ) ) `
+	installedSoftwareStmt = ds.reader(ctx).Rebind(installedSoftwareStmt)
+	var hostInstalledSoftware []*hostSoftware
+	err = sqlx.SelectContext(ctx, ds.reader(ctx), &hostInstalledSoftware, installedSoftwareStmt, args...)
+	if err != nil {
+		return nil, err
 	}
 
-	var excludeVPPAppsClause string
-	if !opts.IsMDMEnrolled {
-		excludeVPPAppsClause = ` AND vat.id IS NULL `
+	return hostInstalledSoftware, nil
+}
+
+func hostSoftwareInstalls(ds *Datastore, ctx context.Context, hostID uint) ([]*hostSoftware, error) {
+	softwareInstallsStmt := `
+        WITH upcoming_software_install AS (
+            SELECT
+                ua.execution_id AS last_install_install_uuid,
+                ua.created_at AS last_install_installed_at,
+                siua.software_installer_id AS installer_id,
+                'pending_install' AS status
+            FROM
+                upcoming_activities ua
+            INNER JOIN
+                software_install_upcoming_activities siua ON ua.id = siua.upcoming_activity_id
+            LEFT JOIN (
+                upcoming_activities ua2
+                INNER JOIN software_install_upcoming_activities siua2 ON ua2.id = siua2.upcoming_activity_id
+            ) ON ua.host_id = ua2.host_id AND
+                siua.software_installer_id = siua2.software_installer_id AND
+                ua.activity_type = ua2.activity_type AND
+                (ua2.priority < ua.priority OR ua2.created_at > ua.created_at)
+            WHERE
+                ua.host_id = :host_id AND
+                ua.activity_type = 'software_install' AND
+                ua2.id IS NULL
+        ),
+        last_software_install AS (
+            SELECT
+                hsi.execution_id AS last_install_install_uuid,
+                hsi.created_at AS last_install_installed_at,
```

Comment:
Probably want `updated_at` here since `hsi.created_at` would be when the install was queued rather than when install status was reported? Same for uninstalls. This works because we filter out `removed = 1` installs and the other changes to rows are ones we want to capture.